pm: compile time device list

gmarull opened the pull request:
Store the list of devices that support device power management at
compile time. If a device uses a function signature other than
`pm_device_none` it will be included in the list. The resulting list is
sorted by the linker by using sections named using init levels and
priorities.

These changes should allow getting rid of most (or al) init code in https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/pm/device.c#L25.

This method requires to not use `NULL` (recently introduced), so please, don't kill me...

gmarull committed:
pm: device: add function signature for devices without device PM

Add "pm_device_none" function to be used when devices to not support
device power management. It is required instead of NULL to make use of
the C preprocessor in future commits.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: store list of devices supporting device PM

Store the list of devices that support device power management at
compile time. If a device uses a function signature other than
`pm_device_none` it will be included in the list. The resulting list is
sorted by the linker by using sections named using init levels and
priorities.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: use pm_device_none

Replace NULL with pm_device_none. This change allows using C
preprocessor at compile time to create a list of devices that support
device power management.

NOTE: incomplete patch, just for testing (hello_world on nrf52840dk).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
[DNM] samples: hello_world: quick test of device PM changes

A quick test, DNM.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

carlescufi commented on include/linker/common-ram.ld:
```diff
@@ -37,6 +37,19 @@
 		__device_end = .;
 	} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
 
+#if CONFIG_PM_DEVICE
```
This should be in `common-rom.ld`?

nashif commented:
> This method requires to not use `NULL` (recently introduced), so please, don't kill me...

:-)

gmarull committed:
pm: device: use compile-time device PM list

The list of devices supporting device PM is now available via
__pm_device_start/end, resolved at compile time.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
Superseded by #35725
